pydantic/_internal/_generate_schema.py

```diff
@@ -75,7 +77,14 @@ def __init__(self, arbitrary_types: bool, types_namespace: dict[str, Any] | None
         self.arbitrary_types = arbitrary_types
         self.types_namespace = types_namespace
 
-    def generate_schema(self, obj: Any) -> core_schema.CoreSchema:  # noqa: C901
+    def generate_schema(self, obj: Any) -> core_schema.CoreSchema:
+        schema = self._generate_schema(obj)
+        modify_json_schema = getattr(obj, '__modify_schema__', None)
+        if modify_json_schema is not None:
+            CoreMetadataHandler(schema).merge_json_schema_misc(JsonSchemaMisc(modify_json_schema=modify_json_schema))
+        return schema
+
+    def _generate_schema(self, obj: Any) -> core_schema.CoreSchema:  # noqa: C901
```

Comment:
@samuelcolvin I wrapped the old generate_schema method (now called _generate_schema) with something that reads the `__modify_schema__` off the input type and updates the metadata as appropriate, regardless of what the input was.

Let me know if you prefer this get handled differently.

Reply:
I'm not sure exactly what the concern is, but I do think the current recursion handles this properly; the following test passes:
```python
def test_modify_schema_dict_keys() -> None:
    class MyType:
        @classmethod
        def __modify_schema__(cls, schema):
            schema['test'] = 'passed'

    class MyModel(BaseModel):
        my_field: MyType

        model_config = dict(arbitrary_types_allowed=True)

    assert MyModel.model_json_schema() == {
        'properties': {'my_field': {'test': 'passed', 'title': 'My Field'}},  # <---- 'test': 'passed'
        'required': ['my_field'],
        'title': 'MyModel',
        'type': 'object',
    }
```